### PR TITLE
[24.1] Fix upload when current history changes

### LINKED ISSUE #18662: Changing current history with a populated uploader prevents users from beginning upload
**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 24.1.2.dev0 (usegalaxy.org)

**To Reproduce**
1. Open the Uploader
2. Add a file to upload (local file or paste)
3. Do not start upload yet, and exit the modal
4. Change your current history
5. Open Uploader again
6. Click Start but nothing happens

**Expected behavior**
Clicking Start should start the upload into this changed current history

**Screencast**

https://github.com/user-attachments/assets/2bac4031-8625-4c3f-92b0-ed6eebbecd7c

### PR DESCRIPTION
**Pull request by davelopez**
Fixes #18662

<del>The upload queue will now be reused if it was not started/running before when you change the target history.</del>

This will ensure that when we change the current history in the middle of the upload each dataset goes to the history where it was scheduled for the first time.

As a bonus, the name of the current history is shown at the top of the upload modal to be more clear about the target history (thanks @ahmedhamidawan for the idea). If this is too much for a fix we can move those commits to dev or drop them if people think is not necessary.

image

## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Follow the steps in #18662

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: client/src/components/Upload/DefaultBox.vue [hunk 1]
```diff
@@ -99,7 +100,16 @@ const listExtensions = computed(() => props.effectiveExtensions.filter((ext) =>
...
 
-const queue = computed(() => createUploadQueue());
+onMounted(() => {
+    queue.value = initUploadQueue();
+});
+
+watch(
+    () => props.historyId,
+    () => {
+        queue.value = initUploadQueue();
```
Comment (mvdbeek) [inline client/src/components/Upload/DefaultBox.vue]: Is this not a problem for uploads that are still running and targeting another history ?

Comment (mvdbeek) [reply client/src/components/Upload/DefaultBox.vue]: I would have assumed you need to associate each upload with a history

Comment (davelopez) [reply client/src/components/Upload/DefaultBox.vue]: We are still associating the whole queue with a history. The only difference is that before we always created a new queue whenever the current history changed, but now we will re-target to the new history only IF the queue was not "started". If it was already running, it will continue running in the old history and we create a "new queue" for the "new history" so whatever we add next is going to be associated with the new one.

So, in summary, once we press "Start" whatever is in the queue will go to the current history if I didn't misunderstand anything.

File: client/src/components/Upload/DefaultBox.vue [hunk 2]
```diff
@@ -99,7 +100,16 @@ const listExtensions = computed(() => props.effectiveExtensions.filter((ext) =>
 const showHelper = computed(() => Object.keys(uploadItems.value).length === 0);
 const uploadValues = computed(() => Object.values(uploadItems.value));
 
-const queue = computed(() => createUploadQueue());
+onMounted(() => {
+    queue.value = initUploadQueue();
+});
```
Comment (ElectronicBlueberry) [inline client/src/components/Upload/DefaultBox.vue]: To avoid duplication, you can put `{ immediate: true }` in the watcher options

Comment (davelopez) [reply client/src/components/Upload/DefaultBox.vue]: I keep forgetting that, thanks!

Comment (ElectronicBlueberry) [reply client/src/components/Upload/DefaultBox.vue]: There's also this neat [shorthand](https://vueuse.org/shared/watchImmediate/)

File: client/src/components/Upload/DefaultBox.vue (hunk 1 above)
Comment (mvdbeek) [reply client/src/components/Upload/DefaultBox.vue]: But you lost the old queue then, how does the progress bar update ?

Comment (davelopez) [reply client/src/components/Upload/DefaultBox.vue]: In my local tests the progress "works". I uploaded 100 files to "history A" and then switched to "history B" in the middle of the upload. The progress keeps going and the files get uploaded to the history A. The only thing is that, if you open the uploader, you need to wait until the progress is completed before you can try to upload anything to "history B". The UI won't let you otherwise.

If this is not the desired behavior I can try to change the implementation to maintain a unique queue and associate the history with each individual item when the user presses "Start" as you suggest.

Comment (mvdbeek) [reply client/src/components/Upload/DefaultBox.vue]: OK, that's fine for 24.1. Yes, I think you should be able to start a new upload at any point, but if that's a larger change targeting dev is fine.

Comment (davelopez) [reply client/src/components/Upload/DefaultBox.vue]: I will keep exploring a bit the option to associate the history with individual items and see if I can make it work without too many changes, if not, we can go with these changes as a fix :+1:

Comment (davelopez) [reply client/src/components/Upload/DefaultBox.vue]: Actually, this is a cleaner and simpler solution! Thanks for suggesting it!

It now lets you pause and schedule upload datasets to the new history and continue. Better user experience I think.

UploadSwithHistoryFix

Comment (ahmedhamidawan) [reply client/src/components/Upload/DefaultBox.vue]: It might just be me but isn't it a little difficult to decipher the difference between what will go to `Old History` and what to `New History`? As in, would it be nicer to have an `<hr>` between the file queues in the uploader?

Comment (davelopez) [reply client/src/components/Upload/DefaultBox.vue]: This is a good point and we should definitely consider how to communicate this better in the future uploader redesign or in dev :+1: 
That said, I would prefer to keep the fix as minimal as possible, I already stretched it a bit by adding the history name at the top :stuck_out_tongue_closed_eyes:

Comment (ahmedhamidawan) [reply client/src/components/Upload/DefaultBox.vue]: Fair point, we can keep the change minimal for now
